### PR TITLE
Select2 4.0 support wip

### PR DESCRIPTION
**Pull request by motin**
JS Bin here: https://jsbin.com/gipezidemi/edit?html,js,output

To make it work with select2 4.0 showed out to be a minor task, thanks to the fact that the underlying element is now a select-element and not an input-element. Most of the directive's code could be removed, and the select2 examples from the readme works as expected.

However, I have not updated any of the tests in the repo (nor will I), and there are probably additional features that were included in the previous angular-select2 that are now broken, so additional work may need to be done. Anyway, it is a start.

https://github.com/rubenv/angular-select2/issues/18

### CODE REVIEW

**Comment by rubenv** [comment]
> However, I have not updated any of the tests in the repo (nor will I)

This is obviously a big blocker for any release. The whole reason we have tests is to make sure we don't break things. We'll need to look at what tests and features are still relevant but simply tossing everything out isn't all that acceptable.

**Comment by motin** [comment]
The development of v4-support is by no means a one-man-show, so while one person does some initial implementation and provides proof-of-concept via jsbin or similar, another one can look at what tests and features are still relevant and yet another can supply updated tests.

File: src/select2.js [hunk 1]
```diff
@@ -244,40 +61,29 @@ angular.module("rt.select2", [])
...
-                                val = optionItems[e.val];
-                                controller.$setViewValue(val ? val.id : null);
-                            }
-
-                            controller.$render();
-
-                        });
-                    });
 
                     element.on("select2-blur", function () {
```
Comment (kevin-brown) [inline src/select2.js]: I might as well mention that there is no longer a `select2-blur` event that is triggered.

**Comment by motin** [comment]
> I've pushed a v4 branch, pull requests are welcome. Once this branch is stable, we'll merge it to master and we'll split off a maintenance branch for the older select2 variant.

According to the above quoted plan, should not this pr be merged so that others can try it out and submit more pull requests to this branch, for instance with test updates, clean-ups and other fixes?

**Comment by rubenv** [comment]
History learns me that if a patch submitter doesn't submit tests, he'll never do it. Nor will anyone else.

I'm all for having a v4 branch, but starting out with no tests at all feels like a very bad idea.

**Comment by motin** [comment]
The test I supplied was in the form of a JS Bin here: https://jsbin.com/gipezidemi/edit?html,js,output

Can we maybe use the same or similar contents/markup/scripts to create tests that are committed in the repo as well?

**Comment by rubenv** [comment]
Shouldn't be too hard to convert the existing Protractor tests to something that works with v4.

**Comment by axar** [comment]
what about initial selection with ajax calls?

**Comment by tonidy** [comment]
@motin any update on this PR? 

cc: @rubenv

**Comment by motin** [comment]
@tonidy Unfortunately not. There are still missing tests, which are mandatory in order to merge the PR. I still suggest that someone else write those tests since I have all but stopped using angular-select2 in my own projects. It would be great to have an answer on the comment from 22 oct as well: https://github.com/rubenv/angular-select2/pull/30#issuecomment-150210207

**Pull request closed without merging**